src/cargo/ops/cargo_compile.rs

```diff
@@ -408,13 +408,24 @@ fn scrape_build_config(config: &Config,
 fn scrape_target_config(config: &Config, triple: &str)
                         -> CargoResult<ops::TargetConfig> {
 
+    let get_path_string = |key: &str| -> CargoResult<Option<String>> {
+        Ok(
+            if let Some(path) = try!(config.get_path(key)) {
+                Some(try!(path.into_os_string().into_string()
+                          .map_err(|_| internal("unhandled non-utf8 path"))))
+            } else {
+                None
+            }
+        )
+    };
+
     let key = format!("target.{}", triple);
-    let ar = try!(config.get_string(&format!("{}.ar", key)));
-    let linker = try!(config.get_string(&format!("{}.linker", key)));
+    let ar = try!(get_path_string(&format!("{}.ar", key)));
+    let linker = try!(get_path_string(&format!("{}.linker", key)));
 
     let mut ret = ops::TargetConfig {
-        ar: ar.map(|p| p.0),
-        linker: linker.map(|p| p.0),
+        ar: ar,
+        linker: linker,
```

Comment:
Ah yeah I think this would just involve pushing `Path` and `PathBuf` through a little into the backend. The `TargetConfig` would store `Option<PathBuf>`, the `linker` and `ar` methods on `Context` would return `&Path` instead of `&str`, the `opt` function would build up with a `OsString` directly and then `build_plugin_args` would be tweaked a bit to handle this. 

I think done the right way you won't have to ever actually return an error! If you need any help with this just lemme know.